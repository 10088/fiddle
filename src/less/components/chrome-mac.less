.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
