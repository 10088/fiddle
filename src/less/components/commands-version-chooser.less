button.version-chooser.bp3-button {
  border-radius: 3px;
}
