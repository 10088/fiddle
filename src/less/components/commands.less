@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
