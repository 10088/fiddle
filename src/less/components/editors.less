.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
