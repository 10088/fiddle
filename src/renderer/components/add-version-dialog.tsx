import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';
import * as semver from 'semver';

import { NpmVersion } from '../../interfaces';
import { getElectronNameForPlatform } from '../../utils/electron-name';
import { AppState } from '../state';
import { Dialog } from './dialog';

export interface AddVersionDialogProps {
  appState: AppState;
}

export interface AddVersionDialogState {
  isValidElectron: boolean;
  isValidVersion: boolean;
  file?: File;
  version: string;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<AddVersionDialogProps, AddVersionDialogState>}
 */
@observer
export class AddVersionDialog extends React.Component<AddVersionDialogProps, AddVersionDialogState> {
  constructor(props: AddVersionDialogProps) {
    super(props);

    this.state = {
      isValidVersion: false,
      isValidElectron: false,
      version: ''
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeFile = this.onChangeFile.bind(this);
    this.onChangeVersion = this.onChangeVersion.bind(this);
  }

  /**
   * Handles a change of the file input
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public async onChangeFile(event: React.ChangeEvent<HTMLInputElement>) {
    const { files } = event.target;
    const { binaryManager } = this.props.appState;
    const file = files && files[0] ? files[0] : undefined;

    const isValidElectron = !!(file
      && file.path
      && await binaryManager.getIsDownloaded('custom', file.path));

    this.setState({ file, isValidElectron });
  }

  public onChangeVersion(event: React.ChangeEvent<HTMLInputElement>) {
    const version = event.target.value || '';
    const isValidVersion = !!semver.valid(version);

    this.setState({
      version,
      isValidVersion,
     });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { file, version } = this.state;

    if (!file) return;

    const name = file.path
      .slice(-20)
      .split(path.sep)
      .slice(1)
      .join(path.sep);

    const toAdd: NpmVersion = {
      localPath: file.path,
      version,
      name
    };

    this.props.appState.addLocalVersion(toAdd);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isAddVersionDialogShowing = false;
    this.reset();
  }

  /**
   * Reset this component's state
   */
  public reset(): void {
    this.setState({
      isValidElectron: false,
      isValidVersion: false,
      version: '',
      file: undefined
    });
  }

  get buttons() {
    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;

    return [
      (
        <button
          className='button'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
        >
          Add
        </button>
      ), (
        <button
          className='cancel'
          key='cancel'
          onClick={this.onClose}
        >
          Cancel
        </button>
      )
    ];
  }

  public render() {
    const { isAddVersionDialogShowing } = this.props.appState;
    const dirOptions = { webkitdirectory: 'true' };

    return (
      <Dialog
        isShowing={isAddVersionDialogShowing}
        isShowingBackdrop={true}
        buttons={this.buttons}
        onClose={this.onClose}
        isCentered={true}
        className='add-version-dialog'
        key='add-version-dialog'
      >
        <label
          htmlFor='custom-electron-version'
          className='force-button'
        >
          Select the folder containing {getElectronNameForPlatform()}
        </label>
        <input
          type='file'
          onChange={this.onChangeFile}
          id='custom-electron-version'
          name='custom-electron-version'
          {...dirOptions}
        />
        {this.renderPath()}
      </Dialog>
    );
  }

  private renderPath(): JSX.Element | null {
    const { file, isValidElectron } = this.state;

    if (!file || !file.path) return null;

    const info = isValidElectron
      ? `We found an ${getElectronNameForPlatform()} in this folder.`
      : `We did not find a ${getElectronNameForPlatform()} in this folder...`;

    return (
      <>
        <input
          readOnly={true}
          value={file.path}
        />
        <p>
          {info}
        </p>
        {this.renderVersionInput()}
      </>
    );
  }

  private renderVersionInput(): JSX.Element | null {
    const { isValidElectron, isValidVersion, version } = this.state;

    if (!isValidElectron) return null;

    return (
      <>
        <p>
          Please specify a version, used for typings and the name.
          Must be <code>semver</code> compliant.
        </p>
        <input
          className={isValidVersion ? '' : 'hasError'}
          value={version}
          onChange={this.onChangeVersion}
          placeholder='4.0.0'
        />
      </>
    );
  }

}
