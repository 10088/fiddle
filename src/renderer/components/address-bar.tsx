import * as React from 'react';
import { observer } from 'mobx-react';
import * as Octokit from '@octokit/rest';
import * as classNames from 'classnames';

import { AppState } from '../state';
import { INDEX_HTML_NAME, MAIN_JS_NAME, RENDERER_JS_NAME } from '../constants';
import { idFromUrl } from '../../utils/gist';
import { reaction } from 'mobx';

export interface AddressBarProps {
  appState: AppState;
}

export interface AddressBarState {
  value: string;
}

@observer
export class AddressBar extends React.Component<AddressBarProps, AddressBarState> {
  constructor(props: AddressBarProps) {
    super(props);

    this.loadFiddle = this.loadFiddle.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: this.props.appState.gistId
    };
  }

  /**
   * Handle the form's submit event, trying to load whatever
   * URL was entered.
   *
   * @param {React.SyntheticEvent<HTMLFormElement>} event
   * @memberof AddressBar
   */
  public async handleSubmit(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();

    this.props.appState.gistId = idFromUrl(this.state.value) || this.state.value;
    if (this.state.value) {
      this.loadFiddle();
    }
  }

  /**
   * Once the component mounts, we'll subscribe to gistId changes
   */
  public componentDidMount() {
    reaction(
      () => this.props.appState.gistId,
      (gistId: string) => this.setState({ value: gistId })
    );
  }

  /**
   * Handle the change event, which usually just updates the address bar's value
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   * @memberof AddressBar
   */
  public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: idFromUrl(event.target.value) || event.target.value });
  }

  /**
   * Load a fiddle
   *
   * @returns
   * @memberof AddressBar
   */
  public async loadFiddle() {
    const { gistId } = this.props.appState;

    if (!confirm('Are you sure you want to load a new fiddle, all current progress will be lost?')) return;

    try {
      const octo = new Octokit();
      const gist = await octo.gists.get({
        gist_id: gistId,
        id: gistId,
      });

      window.ElectronFiddle.app.setValues({
        html: gist.data.files[INDEX_HTML_NAME].content,
        main: gist.data.files[MAIN_JS_NAME].content,
        renderer: gist.data.files[RENDERER_JS_NAME].content,
      });
    } catch (error) {
      console.warn(`Loading fiddle failed`, error);
    }
  }

  public render() {
    const { isUnsaved } = this.props.appState;
    const { value } = this.state;
    const className = classNames('address-bar', isUnsaved, { empty: !value });

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        <span>https://gist.github.com/</span>
        <input
          key='addressbar'
          type='text'
          placeholder='...'
          value={value}
          onChange={this.handleChange}
        />
      </form>
    );
  }
}
