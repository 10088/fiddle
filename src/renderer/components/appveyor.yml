environment:
  matrix:
    - nodejs_version: "8"

init:
  - git config --global core.symlinks true

install:
  - ps: Install-Product node $env:nodejs_version x64
  - node --version
  - npm i

cache:
  - '%APPDATA%\npm-cache -> appveyor.yml'

test_script:
  - npm run lint
  - npm run test

build: off
