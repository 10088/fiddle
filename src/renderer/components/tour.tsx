import * as React from 'react';
import { Dialog } from './dialog';
import { positionForRect, invertPosition } from '../../utils/position-for-rect';

export interface TourScriptStep {
  selector: string;
  content: JSX.Element;
}

export interface TourProps {
  tour: Set<TourScriptStep>;
  onStop: () => void;
}

export interface TourState {
  tour: IterableIterator<Array<TourScriptStep>>;
  step: TourScriptStep | null;
}

export class Tour extends React.Component<TourProps, TourState> {
  private resizeHandle = 0;

  constructor(props: TourProps) {
    super(props);

    this.advance = this.advance.bind(this);
    this.stop = this.stop.bind(this);
    this.onResize = this.onResize.bind(this);

    this.state = {
      tour: props.tour.entries(),
      step: null,
    };
  }

  /**
   * Handles a resize of the window.
   */
  public onResize() {
    if (!this.resizeHandle) {
      this.resizeHandle = window.setTimeout(() => {
        this.forceUpdate();
        this.resizeHandle = 0;
      }, 100);
    }
  }

  /**
   * Component was mounted: Sign up for the window's "resize"
   * event and move to the first step of the tour.
   */
  public componentDidMount() {
    this.advance();

    window.addEventListener('resize', this.onResize);
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  /**
   * Moves the tour to the next step
   */
  public advance() {
    const { done, value } = this.state.tour.next();
    const step = done ? null : value[0];

    this.setState({ step });
  }

  /**
   * Stops the tour
   */
  public stop() {
    this.props.onStop();
  }

  public render() {
    const { step } = this.state;

    if (!step) return null;

    return (
      <div className='tour'>
        {this.getStep(step)}
      </div>
    );
  }

  private getDialogForStep({ content }: TourScriptStep, rect: ClientRect) {
    const buttons = [
      <button key='btn-stop' onClick={this.stop}>Stop Tour</button>,
      <button key='btn-adv' onClick={this.advance}>Continue</button>
    ];

    const size = { width: 300, height: 300 };
    const position = positionForRect(rect, size);
    const style: React.CSSProperties = {
      position: 'absolute',
      top: position.top,
      left: position.left,
      width: size.width
    };
    const arrow = invertPosition(position.type);

    return (
      <Dialog buttons={buttons} style={style} arrow={arrow}>
        {content}
      </Dialog>
    );
  }

  /**
   * Returns the "mask" for a given step.
   *
   * @param {TourScriptStep} { selector, text }
   * @returns {(JSX.Element | null)}
   */
  private getStep(step: TourScriptStep): JSX.Element | null {
    const { selector } = step;
    const element = document.querySelector(selector);
    if (!element) return null;

    const rect = element.getBoundingClientRect();
    const { top, left, width, height } = rect;

    return (
      <>
        {this.getDialogForStep(step, rect)}
        <svg height='100%' width='100%'>
          <rect fill='rgba(0, 0, 0, 0.5)' x='0' y='0' mask='url(#mask)' height='100%' width='100%'/>
          <mask id='mask' maskUnits='userSpaceOnUse' maskContentUnits='userSpaceOnUse'>
            <rect className='bg' x='0' y='0' fill='white' height='100%' width='100%' />
            <rect x={left} y={top} width={width} height={height} fill='black' rx='5' ry='5' />
          </mask>
        </svg>
      </>
    );
  }
}
