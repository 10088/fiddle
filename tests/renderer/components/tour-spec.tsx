import { shallow } from 'enzyme';
import * as React from 'react';

import { Tour } from '../../../src/renderer/components/tour';
import { overridePlatform, resetPlatform } from '../../utils';

describe('VersionChooser component', () => {
  const oldQuerySelector = document.querySelector;
  const mockTour = new Set([
    {
      name: 'mock-step-1',
      selector: 'div.mock-1',
      content: (
        <span key='1'>mock-step-1</span>
      )
    },
    {
      name: 'mock-step-2',
      selector: 'div.mock-2',
      content: (
        <span key='2'>mock-step-2</span>
      )
    }
  ]);

  beforeAll(() => {
    overridePlatform('darwin');
  });

  afterAll(() => {
    resetPlatform();
  });

  beforeEach(() => {
    document.querySelector = jest.fn(() => ({
      getBoundingClientRect: jest.fn(() => ({
        top: 20, left: 25, height: 120, width: 130
      }))
    }));
  });

  afterEach(() => {
    document.querySelector = oldQuerySelector;
  });

  it('renders', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders custom buttons', () => {
    const mockStop = jest.fn();
    const btnTour = new Set([
      {
        name: 'mock-step-1',
        selector: 'div.mock-1',
        content: (
          <span>mock-step-1</span>
        ),
        getButtons: () => [ (<span key='1'>hi</span>) ]
      }
    ]);

    const wrapper = shallow(<Tour tour={btnTour} onStop={mockStop} />);
    expect(wrapper.html().includes('<span>hi</span>')).toBe(true);
  });

  it('stops on stop()', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.stop();

    expect(mockStop).toHaveBeenCalled();
  });

  it('advances on advance()', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.advance();
    expect((wrapper.state('step') as any).name).toBe('mock-step-2');

    instance.advance();
    expect(wrapper.state('step')).toBe(null);
  });

  it('handles a resize', () => {
    const oldTimeout = window.setTimeout;
    window.setTimeout = (fn: any) => {
      fn();
      return 1;
    };

    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.forceUpdate = jest.fn();
    instance.onResize();

    expect(instance.resizeHandle).toBeTruthy();
    expect(instance.forceUpdate).toHaveBeenCalled();

    window.setTimeout = oldTimeout;
  });
});
