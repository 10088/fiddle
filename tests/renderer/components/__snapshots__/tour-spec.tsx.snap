// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionChooser component renders 1`] = `
<div
  className="tour"
>
  <Dialog
    arrow="left"
    buttons={
      Array [
        <button
          className="btn-stop"
          onClick={[Function]}
        >
          Stop Tour
        </button>,
        <button
          className="btn-adv"
          onClick={[Function]}
        >
          Continue
        </button>,
      ]
    }
    isCentered={false}
    isShowing={true}
    isShowingBackdrop={false}
    key="mock-step-1"
    style={
      Object {
        "left": 165,
        "margin": "10px",
        "position": "absolute",
        "top": 5,
        "width": 390,
      }
    }
  >
    <span
      key="1"
    >
      mock-step-1
    </span>
  </Dialog>
  <svg
    height="100%"
    width="100%"
  >
    <rect
      fill="rgba(0, 0, 0, 0.65)"
      height="100%"
      mask="url(#mask)"
      width="100%"
      x="0"
      y="0"
    />
    <mask
      id="mask"
      maskContentUnits="userSpaceOnUse"
      maskUnits="userSpaceOnUse"
    >
      <rect
        className="bg"
        fill="white"
        height="100%"
        width="100%"
        x="0"
        y="0"
      />
      <rect
        fill="black"
        height={120}
        rx="5"
        ry="5"
        stroke="black"
        strokeWidth="5"
        width={130}
        x={25}
        y={20}
      />
    </mask>
  </svg>
</div>
`;
